hermes-client/src/main/java/pl/allegro/tech/hermes/client/ReactiveHermesClient.java

```diff
@@ -0,0 +1,324 @@
+        return publish(hermesMessage(topic, message).json().build());
+    }
+
+    public Mono<HermesResponse> publishJSON(String topic, String message) {
+        return publish(hermesMessage(topic, message).json().build());
+    }
+
+    public Mono<HermesResponse> publishAvro(String topic, int schemaVersion, byte[] message) {
+        return publish(hermesMessage(topic, message).avro(schemaVersion).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String message) {
+        return publish(hermesMessage(topic, message).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String contentType, byte[] message) {
+        return publish(hermesMessage(topic, message).withContentType(contentType).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String contentType, String message) {
+        return publish(hermesMessage(topic, message).withContentType(contentType).build());
+    }
+
+    public Mono<HermesResponse> publish(String topic, String contentType, int schemaVersion, byte[] message) {
+        return publish(hermesMessage(topic, message).withContentType(contentType).withSchemaVersion(schemaVersion).build());
+    }
+
+    public Mono<HermesResponse> publish(HermesMessage message) {
+        if (shutdown) {
+            return completedWithShutdownException();
+        }
+        HermesMessage.appendDefaults(message, defaultHeaders);
+        return publishWithRetries(message);
+    }
+
+    public boolean addMessageDeliveryListener(MessageDeliveryListener listener) {
+        return messageDeliveryListeners.add(listener);
+    }
+
+    private Mono<HermesResponse> publishWithRetries(HermesMessage message) {
+        currentlySending.incrementAndGet();
+        Mono<Result> sendOnceResult = sendOnce(message)
+                .flatMap(this::testRetryCondition)
+                .onErrorResume(this::mapExceptionToFailedAttempt)
+                .subscribeOn(scheduler);
+
+        return retry(message, sendOnceResult)
+                .doOnSuccess(hr -> {
+                    if (hr.response.isSuccess()) {
+                        handleSuccessfulRetry(message, hr.getAttempt());
+                    } else {
+                        handleFailure(message, hr.getAttempt());
+                    }
+                })
+                .map(result -> result.response)
+                .doFinally(s -> currentlySending.decrementAndGet());
+    }
+
+    private Mono<Attempt> retry(HermesMessage message, Mono<Result> sendOnceResult) {
+        Retry retrySpec = prepareRetrySpec(message);
+        return sendOnceResult
+                .flatMap(this::unwrapFailedAttemptAsException)
+                .retryWhen(retrySpec)
+                .subscriberContext(ctx -> ctx.put(RETRY_CONTEXT_KEY, HermesRetryContext.emptyRetryContext()))
+                .onErrorResume(Exception.class, exception -> {
+                    if (isRetryExhausted(exception)) {
```

Comment:
Rest of the functions used here are extracted, why isn't this one as well?

Reply:
There's no particular reason. I extracted a method.